Add aarch64 support

rickvantooster opened the pull request.

rickvantooster committed:
Added aarch64 support

rickvantooster committed:
Fixed wrong SM in sleep_write

0dminnimda commented:
Nice! That works on termux aarch64, except their gcc reports an error for `(void)fd`, a guard will fix it:
```c
#ifndef __ANDROID__
    (void) fd;
#endif
```

---

I really don't know arm, but this (https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#611general-purpose-registers) document on aarch64 suggests that:

> A subroutine invocation must preserve the contents of the registers r19-r29 and SP. All 64 bits of each value stored in r19-r29 must be preserved.

And for SIMD/floating-point:

>  Registers v8-v15 must be preserved by a callee across subroutine calls; [...]. Additionally, only the bottom 64 bits of each value stored in v8-v15 need to be preserved; it is the responsibility of the caller to preserve larger values.

And if I read and understood the code correctly, you are saving r19-r30, but what about v8-v15?

rickvantooster committed:
Added guard around fd cast and added code to save the v8-v15 registers

rickvantooster commented:
> And if I read and understood the code correctly, you are saving r19-r30, but what about v8-v15?

Thank you for pointing it out. They are now being saved. 
I've also added the guard around the fd void cast. However shouldn't
 ```
(void)rsp;
``` 
in coroutine_restore_context also have the same guard?

I also added the inline assembly to move fd to the correct registers in sleep_write and sleep_read as I forgot to add those in my initial commit 🥲.

0dminnimda commented:
> in coroutine_restore_context also have the same guard?

yes, all of the cases with `(void)...` inside naked functions. It's like with clang, compilers don't want to see any code except inline assembly inside naked functions.

0dminnimda commented on coroutine.c:
```diff
@@ -251,6 +339,44 @@ void coroutine_go(void (*f)(void*), void *arg)
+
+
+
+    *(--rsp) = 0;   // push r29
```
Don't you also need to reserve space for r30 here, since the load will read both r29 and r30?

rickvantooster replied:
> Don't you also need to reserve space for r30 here, since the load will read both r29 and r30?

r30 is the same as the link register so the line
```
*(--rsp) = coroutine__finish_current;
``` 
will already reserve space for it.

0dminnimda replied:
Ok, yeah, we don't need to preserve it's value, so we can just read some garbage into it, instead of the previous value. Now but what about storing to the stack, wouldn't it override `*(--rsp) = f; // push r0`?

rickvantooster replied:
I am not sure if I am following you. So far everything runs perfectly fine on the raspberry pi 4 model B.
